### PR TITLE
Loop Refactor 5/N - Prediction Loop

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

This was originally the full Loop Refactor POC PR which got split up into multiple parts #7871, #7985, #7990, #8056  and what remains now is the prediction loop refactor.

This PR now only contains the Prediction Loop refactors. There are two new classes based on the existing interface in master:
- `PredictionDataLoaderLoop`: Loops over all dataloaders and for each runs a `PredictionEpochLoop`
- `PredictionEpochLoop`: Loops over a single dataloader of prediction batches

After merge:
- Follow up on TODOs in #7938 

- Related PRs for review: #8024 , #8025 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

co-authored by @justusschock

### CODE REVIEW

**Comment by Borda** [comment]
cool branch name `refactor/loops_everywhere` :D

File: pytorch_lightning/trainer/properties.py [hunk 1]
```diff
@@ -477,6 +480,19 @@ def sanity_checking(self, val: bool) -> None:
     Loop properties
     """
 
+    @property
+    def epoch_loop(self) -> EpochLoop:
+        # TODO: the current train_loop should be renamed to epoch_loop
+        return self.train_loop
```
Comment (justusschock) [inline pytorch_lightning/trainer/properties.py:486]: maybe make the attribute protected and deprecate the self.train_loop property? May be confusing to get the difference between train_loop and training_loop

**Comment by carmocca** [comment]
Should the new loops save/load state? Or should they always be fully re-created?

**Comment by awaelchli** [comment]
For checkpointing yes, they could have a state dict for the checkpoint. Apart from that I think we can simply rely on the implementation of `Loop.reset()` when multiple calls to `Loop.run()` are made.

**Comment by carmocca** [comment]
Remember to update the CODEOWNERS for the new loops directory

**Comment by carmocca** [comment]
Is there a document or snippet describing the new loop hierarchy? for fit, eval and predict.

Which ones do we have and what holds what

File: pytorch_lightning/loops/base.py [hunk 2]
```diff
@@ -59,7 +59,8 @@ def skip(self) -> bool:
 
     def connect(self, trainer: 'pl.Trainer', *args: Any, **kwargs: Any) -> None:
         """Connects Loop with all the necessary things like connectors and accelerators."""
-        self.trainer = proxy(trainer)
+        # TODO(@justusschock): Make the trainer a weakref/proxy
+        self.trainer = trainer
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/loops/base.py:62]: I will do this in a separate PR since it touches all loops.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/prediction_writer.py [hunk 3]
```diff
@@ -109,7 +109,7 @@ def on_predict_batch_end(
         if not self.interval.on_batch:
             return
         is_distributed = trainer.accelerator_connector.is_distributed
-        batch_indices = trainer.predict_loop.batch_indices if is_distributed else None
+        batch_indices = trainer.predict_loop.prediction_loop.current_batch_indices if is_distributed else None
```
Comment (carmocca) [inline pytorch_lightning/callbacks/prediction_writer.py:112]: `predict_loop.prediction_loop`

is confusing. Can we change the variable names?

`predict_loop.batch_loop`?

Comment (awaelchli) [reply pytorch_lightning/callbacks/prediction_writer.py:112]: there is no batch loop for prediction (currently). remember: x_loop means it loops over one instance of x, e.g. epoch_loop runs over a single epoch. 

`predict_loop.epoch_loop?`

Comment (carmocca) [reply pytorch_lightning/callbacks/prediction_writer.py:112]: sure

File: pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+import pytorch_lightning as pl
+from pytorch_lightning.loops.dataloader.dataloader_loop import DataLoaderLoop
+from pytorch_lightning.loops.prediction_loop import PredictionLoop
+from pytorch_lightning.plugins import DDPSpawnPlugin
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.types import _PREDICT_OUTPUT
+
+
+class PredictionDataLoaderLoop(DataLoaderLoop):
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py:14]: I think it'd look nicer if this was just called `PredictLoop`.

IMO the fact that it subclasses `DatLoader` loop is a detail.

Just as your `class Square(Rectangle)` wouldn't be named `SquareRectangle`

This also applies to the existing loops

Comment (awaelchli) [reply pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py:14]: sounds good, let me rename all loops together  in a new PR where we can also decide on the package structure how we want to organize files?

Comment (carmocca) [reply pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py:14]: Sure

File: pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        self.epoch_loop.connect(trainer, *args, **kwargs)
+
+    def reset(self) -> None:
+        """Resets the internal state of the loop for a new run"""
+        super().reset()
+        self.predictions = []
+        self.epoch_batch_indices = []
+
+    def on_run_start(self) -> None:
+        """Calls ``on_predict_start`` hook"""
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py:87]: Personally not a fan of docstrings which just enumerate the calls in the code. But no strong feeling

```suggestion
```

Same comment for several other docstrings in this file

File: pytorch_lightning/loops/prediction_epoch_loop.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        self._all_batch_indices: List[int] = []
+
+    @property
+    def done(self) -> bool:
+        """Ends prediction when the iteration count exceeds the total number of available batches"""
+        return self.iteration_count >= self._dl_max_batches
+
+    @property
+    def should_store_predictions(self) -> bool:
+        """whether the predictions should be stored for later usage (e.g. aggregation or returning)"""
```
Comment (carmocca) [inline pytorch_lightning/loops/prediction_epoch_loop.py:31]:
```suggestion
        ""Whether the predictions should be stored for later usage (e.g. aggregation or returning)"""
```

File: pytorch_lightning/loops/prediction_epoch_loop.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    def on_run_start(
+        self,
+        dataloader_iter: Iterator,
+        dataloader_idx: int,
+        dl_max_batches: int,
+        num_dataloaders: int,
+        return_predictions: bool = False
+    ) -> None:
+        """Prepares the loops internal state
```
Comment (carmocca) [inline pytorch_lightning/loops/prediction_epoch_loop.py:49]:
```suggestion
        """
        Prepares the loops internal state
```

File: pytorch_lightning/loops/prediction_epoch_loop.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+        Args:
+            dataloader_iter: the iterator over the current dataloader
+            dataloader_idx: the index of the current dataloader
+            dl_max_batches: the maximum number of batches the current loader can produce
+            num_dataloaders: the total number of dataloaders
+            return_predictions: whether to return the obtained predictions
+        """
+        void(dataloader_iter, dataloader_idx)
```
Comment (carmocca) [inline pytorch_lightning/loops/prediction_epoch_loop.py:58]: Should all these arguments be passed with keywords so they can be hidden with `kwargs`?

Otherwise, should we not include them in the docstring?

File: pytorch_lightning/loops/prediction_epoch_loop.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        self._dl_max_batches = dl_max_batches
+        self._num_dataloaders = num_dataloaders
+        self.return_predictions = return_predictions
+
+    def advance(
+        self, dataloader_iter: Iterator, dataloader_idx: int, dl_max_batches: int, *args: Any, **kwargs: Any
+    ) -> None:
+        """Runs one prediction step.
```
Comment (carmocca) [inline pytorch_lightning/loops/prediction_epoch_loop.py:66]:
```suggestion
        """
        Runs one prediction step.

```

File: pytorch_lightning/loops/prediction_epoch_loop.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        if predictions is None:
+            self._warning_cache.warn("predict returned None if it was on purpose, ignore this warning...")
+
+        self.trainer.call_hook("on_predict_batch_end", predictions, batch, batch_idx, dataloader_idx)
+
+        if self.should_store_predictions:
+            self.predictions.append(predictions)
+
+    def _build_kwargs(self, batch: Any, batch_idx: int, dataloader_idx: int) -> Dict[str, Any]:
+        """Assembles the keyword arguments for the ``predict_step``
```
Comment (carmocca) [inline pytorch_lightning/loops/prediction_epoch_loop.py:122]:
```suggestion
        """
        Assembles the keyword arguments for the ``predict_step``
```

File: pytorch_lightning/loops/prediction_epoch_loop.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    def advance(
+        self, dataloader_iter: Iterator, dataloader_idx: int, dl_max_batches: int, *args: Any, **kwargs: Any
+    ) -> None:
+        """Runs one prediction step.
+        Args:
+            dataloader_iter: the iterator over the current dataloader
+            dataloader_idx: the index of the current dataloader
+            dl_max_batches: the maximum number of batches the current loader can produce
+            num_dataloaders: the total number of dataloaders
+            return_predictions: whether to return the obtained predictions
```
Comment (carmocca) [inline pytorch_lightning/loops/prediction_epoch_loop.py:72]: Not in the signature

```suggestion
```

File: pytorch_lightning/loops/prediction_epoch_loop.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: @justusschock what was your intention here? could do either

Comment (justusschock) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: I think all the arguments that are not currently used can become **kwargs.

For removal of the docstrings: I have mixed feelings on removing them: on the one hand, they are currently unused, but on the other hand when we might open this part to user-API, I think it might be nice for them to know what the passed objects are.

I'm fine with either way.

File: pytorch_lightning/loops/prediction_epoch_loop.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/loops/prediction_epoch_loop.py:31]: you deleted a quotation mark, addressing separately.

File: pytorch_lightning/loops/prediction_epoch_loop.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: ok, I vote for keeping the args for now, we don't expose the docs yet in the api reference, so we may re-evaluate this when we do?

Comment (carmocca) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: >  I vote for keeping the args for now

I'm fine with this if they become kwargs

Comment (awaelchli) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: I wasn't sure what you wanted me to do
https://github.com/PyTorchLightning/pytorch-lightning/pull/7700/commits/2639c784a0ff68ac8d8e649cacf8d8195990fb87

File: pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py:14]: #8077

File: pytorch_lightning/callbacks/prediction_writer.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/prediction_writer.py]: renamed epoch loop
also: #8077

File: pytorch_lightning/loops/prediction_epoch_loop.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: I meant doing:

```python
    def on_run_start(
        self,
        dl_max_batches: int,
        num_dataloaders: int,
        return_predictions: bool = False,
        **kwargs: Any,  # dataloader_iter, dataloader_idx
    ) -> None:
```

Comment (awaelchli) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: but `dataloader_iter, dataloader_idx` are supposed to be mandatory and documented.
I am highly confused by this, apologies.

Comment (carmocca) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: > are supposed to be mandatory and documented.

My point is that the fact that we "void" them suggests me that they should not be mandatory (aka passed as kwargs)

Comment (awaelchli) [reply pytorch_lightning/loops/prediction_epoch_loop.py:58]: the reviews recommended this, as these were unused variables and that's apparently bad. this is also why a call to void() was added. 

these args apart from `return_predictions` are definitely mandatory and do not have defaults at this moment.

File: pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    def __init__(self):
+        super().__init__()
+        self.epoch_loop: PredictionEpochLoop = PredictionEpochLoop()
+        self.predictions: Optional[List[List[Any]]] = None
+        self.epoch_batch_indices: Optional[List[List[int]]] = None
+        self._return_predictions: bool = False
+
+    @property
+    def return_predictions(self) -> bool:
+        """whether to return the predictions or not"""
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py:26]:
```suggestion
        """Whether to return the predictions or not"""
```

File: pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        return self._return_predictions
+
+    @return_predictions.setter
+    def return_predictions(self, return_predictions: Optional[bool] = None) -> None:
+        """whether to return the predictions or not"""
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/prediction_dataloader_loop.py:31]: Setter needs docstring?

```suggestion
```

**Review by carmocca — APPROVED** [review]

File: setup.cfg [hunk 14]
```diff
@@ -45,6 +45,8 @@ exclude_lines =
 omit =
     pytorch_lightning/cluster_environments/*.py
     pytorch_lightning/utilities/distributed.py
+    pytorch_lightning/trainer/evaluation_loop.py
+    pytorch_lightning/trainer/predict_loop.py
```
Comment (Borda) [inline setup.cfg:49]: there shall be some TODO, right?

**Review by Borda — APPROVED** [review]

File: setup.cfg (hunk 14 above)
Comment (awaelchli) [reply setup.cfg:49]: I have the pr to remove this ready together with old loop removal, will send right now.

**Review by tchaton — APPROVED** [review]
LGTM !
